File: vps-telegram-bot/vpsupload_bot/video_bot_mod/ui.py
"""Telegram 菜单、分页与确认界面。"""

import logging
import math
import os
from typing import List

from telegram import InlineKeyboardButton, InlineKeyboardMarkup, Update
from telegram.ext import ContextTypes

from .auth import is_admin
from .config import BASE_DIR, ITEMS_PER_PAGE, VIDEO_EXTENSIONS
from .media_utils import assert_path_inside_base, get_formatted_file_size, safe_join
from .task_manager import get_active_task, get_youtube_upload_count

logger = logging.getLogger(__name__)

async def render_main_menu(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if not is_admin(update):
        return

    keyboard = [
        [InlineKeyboardButton("📂 浏览远程文件", callback_data="init_browse")],
        [
            InlineKeyboardButton("📡 RTMP 单路推流", callback_data="init_stream"),
            InlineKeyboardButton("☁️ YouTube 上传", callback_data="init_youtube"),
        ],
        [InlineKeyboardButton("✂️ 智能视频合并", callback_data="init_concat")],
        [InlineKeyboardButton("🔄 批量转码 MP4", callback_data="init_convert")],
        [InlineKeyboardButton("🗑️ 批量删除文件", callback_data="init_delete")],
    ]

    active_name = context.user_data.get("active_task_name") if get_active_task(context) else None
    busy_text = f"\n🔒 当前独占任务: `{active_name}`" if active_name else ""
    upload_count = get_youtube_upload_count(context)
    upload_text = f"\n☁️ YouTube 上传/排队: `{upload_count}`" if upload_count else ""
    text = (
        f"=== 🎬 VPS 多媒体主控面板 ===\n"
        f"根目录: `{BASE_DIR}`\n"
        f"💡 提示: /uploads 查看上传，/stop 中断运行任务"
        f"{busy_text}"
        f"{upload_text}"
    )

    if update.callback_query:
        await update.callback_query.edit_message_text(text, reply_markup=InlineKeyboardMarkup(keyboard), parse_mode="Markdown")
    else:
        await update.message.reply_text(text, reply_markup=InlineKeyboardMarkup(keyboard), parse_mode="Markdown")

async def render_file_selector(update: Update, context: ContextTypes.DEFAULT_TYPE, action_type: str, page: int):
    query = update.callback_query

    current_dir = context.user_data.get("current_dir", BASE_DIR)
    try:
        current_dir = assert_path_inside_base(current_dir)
    except ValueError:
        current_dir = BASE_DIR
        context.user_data["current_dir"] = BASE_DIR

    if not os.path.exists(current_dir):
        await query.edit_message_text("❌ 目录不存在！")
        return

    all_items = os.listdir(current_dir)
    dirs = []
    files = []
    for item in all_items:
        try:
            item_path = safe_join(current_dir, item)
        except ValueError:
            logger.warning("跳过越界路径: %s", item)
            continue

        if os.path.isdir(item_path):
            dirs.append(item)
        elif os.path.isfile(item_path) and item.lower().endswith(VIDEO_EXTENSIONS):
            files.append(item)

    dirs = sorted(dirs)
    files = sorted(files)

    items = dirs + files
    context.user_data["current_files"] = items

    is_multi_select = action_type in ["youtube", "concat", "convert", "delete"]
    selected_indices = context.user_data.setdefault(f"selected_{action_type}", set())

    total_pages = max(1, math.ceil(len(items) / ITEMS_PER_PAGE))
    page = max(0, min(page, total_pages - 1))
    start_idx = page * ITEMS_PER_PAGE
    current_page_items = items[start_idx:start_idx + ITEMS_PER_PAGE]

    keyboard = []
    for i, item_name in enumerate(current_page_items):
        real_idx = start_idx + i
        item_path = safe_join(current_dir, item_name)

        if os.path.isdir(item_path):
            btn_text = f"📁 {item_name}"
            callback_data = f"enterdir_{action_type}_{real_idx}"
        else:
            size_str = get_formatted_file_size(item_path)
            if is_multi_select:
                checkbox = "✅ " if real_idx in selected_indices else "⬜️ "
                btn_text = f"{checkbox}[{size_str}] {item_name}"
                callback_data = f"toggle_{action_type}_{real_idx}_{page}"
            else:
                btn_text = f"[{size_str}] {item_name}"
                callback_data = f"execsingle_{action_type}_{real_idx}"

        keyboard.append([InlineKeyboardButton(btn_text, callback_data=callback_data)])

    nav_buttons = []
    if page > 0:
        nav_buttons.append(InlineKeyboardButton("⬅️ 上一页", callback_data=f"menu_{action_type}_{page - 1}"))
    if page < total_pages - 1:
        nav_buttons.append(InlineKeyboardButton("➡️ 下一页", callback_data=f"menu_{action_type}_{page + 1}"))
    if nav_buttons:
        keyboard.append(nav_buttons)

    if os.path.realpath(current_dir) != os.path.realpath(BASE_DIR):
        keyboard.append([InlineKeyboardButton("⬆️ 返回上一级目录", callback_data=f"updir_{action_type}")])

    if is_multi_select and selected_indices:
        keyboard.append([InlineKeyboardButton(f"▶️ 确认执行 ({len(selected_indices)} 个文件)", callback_data=f"execbatch_{action_type}")])

    keyboard.append([InlineKeyboardButton("🔙 返回主菜单", callback_data="menu_main")])

    action_name_map = {
        "browse": "浏览与查看详情",
        "stream": "推流",
        "youtube": "上传 YT",
        "concat": "合并",
        "convert": "转码 MP4",
        "delete": "删除",
    }

    rel_path = os.path.relpath(current_dir, BASE_DIR)
    display_path = "🏠" if rel_path == "." else f"🏠/{rel_path}"
    header = f"📂 路径: `{display_path}`\n👉 模式: [{action_name_map.get(action_type, action_type.upper())}] (页 {page + 1}/{total_pages})"

    if get_active_task(context):
        header += f"\n🔒 独占任务: `{context.user_data.get('active_task_name', '任务')}`"

    upload_count = get_youtube_upload_count(context)
    if upload_count:
        header += f"\n☁️ YouTube 上传/排队: `{upload_count}`"

    if not items:
        header += "\n\n⚠️ 当前目录下既无子文件夹也无视频文件。"

    await query.edit_message_text(header, reply_markup=InlineKeyboardMarkup(keyboard), parse_mode="Markdown")

async def render_delete_confirmation(update: Update, context: ContextTypes.DEFAULT_TYPE, target_files: List[str]):
    query = update.callback_query
    context.user_data["pending_delete_files"] = target_files

    preview_count = 12
    preview_lines = [f"• `{os.path.basename(path)}`" for path in target_files[:preview_count]]
    if len(target_files) > preview_count:
        preview_lines.append(f"... 其余 {len(target_files) - preview_count} 个文件未展开")

    keyboard = [
        [
            InlineKeyboardButton("✅ 确认删除", callback_data="confirm_delete"),
            InlineKeyboardButton("取消", callback_data="cancel_delete"),
        ]
    ]
    await query.edit_message_text(
        "⚠️ **二次确认：即将永久删除以下文件**\n\n"
        + "\n".join(preview_lines)
        + "\n\n删除后不可恢复。",
        reply_markup=InlineKeyboardMarkup(keyboard),
        parse_mode="Markdown",
    )

File: vps-telegram-bot/vpsupload_bot/RUST/src/ui.rs
use std::sync::Arc;
use std::path::Path;
use teloxide::types::{InlineKeyboardButton, InlineKeyboardMarkup};
use crate::state::AppState;
use crate::media_utils::*;

fn escape_html(input: &str) -> String {
    input.replace('&', "&amp;")
         .replace('<', "&lt;")
         .replace('>', "&gt;")
}

pub async fn build_main_menu(state: Arc<AppState>, user_id: i64) -> (String, InlineKeyboardMarkup) {
    let session = state.get_session(user_id).await;
    let active_guard = state.active_task.lock().await;
    
    let busy_text = match &*active_guard {
        Some(t) => format!("\n⚡ <b>[当前任务]</b> <code>{}</code>", escape_html(&t.name)),
        None => "\n🟢 <b>[系统状态]</b> <code>空闲中 (随时就绪)</code>".to_string(),
    };

    let upload_count = state.youtube_uploads.lock().await.len();
    let upload_text = if upload_count > 0 {
        format!("\n☁️ <b>[上传任务]</b> <code>{} 个任务在排队/运行</code>", upload_count)
    } else {
        String::new()
    };

    let theme_name = match session.progress_bar_theme {
        1 => "🟩🟩⬜⬜ 彩色水果",
        2 => "━━── 简约细线",
        _ => "▰▰▱▱ 科幻方块",
    };

    let text = format!(
        "<b>╔══════ 🎬 VPS 多媒体控制面板 ══════╗</b>\n\n\
         📁 <b>存储根目录:</b> <code>{}</code>\n\
         🎨 <b>当前进度条皮肤:</b> <code>{}</code>\n\
         {}\n\
         {}\n\n\
         <i>💡 提示: 发送 /uploads 查看上传，/stop 中断独占任务。</i>\n\
         <b>╚════════════════════════════════╝</b>",
         escape_html(&state.config.base_dir.to_string_lossy()),
         theme_name,
         busy_text,
         upload_text
    );

    let keyboard = InlineKeyboardMarkup::new(vec![
        vec![InlineKeyboardButton::callback("📂 浏览与操作本地文件", "init_browse")],
        vec![
            InlineKeyboardButton::callback("📡 RTMP 推流", "init_stream"),
            InlineKeyboardButton::callback("☁️ YouTube 上传", "init_youtube"),
        ],
        vec![
            InlineKeyboardButton::callback("✂️ 智能视频合并", "init_concat"),
            InlineKeyboardButton::callback("🔄 批量转码 MP4", "init_convert"),
        ],
        vec![
            InlineKeyboardButton::callback("🎨 切换进度条皮肤", "menu_skin_settings"),
            InlineKeyboardButton::callback("🗑️ 批量清理文件", "init_delete"),
        ],
    ]);

    (text, keyboard)
}

pub fn build_skin_selector_menu() -> (String, InlineKeyboardMarkup) {
    let text = "<b>🎨 选择你喜爱的进度条“视觉皮肤”：</b>\n\n\
                设置后，推流、转码、以及 YouTube 上传进度条将实时切换为该样式。".to_string();

    let keyboard = InlineKeyboardMarkup::new(vec![
        vec![InlineKeyboardButton::callback("▰▰▱▱ 科幻方块", "set_skin_0")],
        vec![InlineKeyboardButton::callback("🟩🟩⬜⬜ 彩色水果", "set_skin_1")],
        vec![InlineKeyboardButton::callback("━━── 简约细线", "set_skin_2")],
        vec![InlineKeyboardButton::callback("🔙 返回主菜单", "menu_main")],
    ]);

    (text, keyboard)
}

pub async fn build_file_selector(
    state: Arc<AppState>,
    user_id: i64,
    action_type: &str,
    page: usize,
) -> (String, InlineKeyboardMarkup) {
    let mut session = state.get_session(user_id).await;
    let current_dir = session.current_dir.clone();

    let mut dirs = Vec::new();
    let mut files = Vec::new();

    if let Ok(entries) = std::fs::read_dir(&current_dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            let name = path.file_name().unwrap().to_string_lossy().to_string();
            if path.is_dir() {
                dirs.push(name);
            } else if path.is_file() {
                let ext = path.extension().and_then(|s| s.to_str()).unwrap_or("").to_lowercase();
                if state.config.video_extensions.iter().any(|e| e.ends_with(&ext)) {
                    files.push(name);
                }
            }
        }
    }
    dirs.sort();
    files.sort();

    let mut all_items = dirs.clone();
    all_items.extend(files.clone());
    session.current_files = all_items.clone();
    state.save_session(user_id, session.clone()).await;

    let is_multi = matches!(action_type, "youtube" | "concat" | "convert" | "delete");
    let selected_set = session.get_selected(action_type).clone();

    let per_page = state.config.items_per_page;
    let total_pages = (all_items.len() + per_page - 1) / per_page;
    let page = page.min(total_pages.saturating_sub(1));
    let start_idx = page * per_page;
    let end_idx = (start_idx + per_page).min(all_items.len());

    let mut rows = Vec::new();
    for i in start_idx..end_idx {
        let name = &all_items[i];
        let path = current_dir.join(name);
        
        if path.is_dir() {
            rows.push(vec![InlineKeyboardButton::callback(
                format!("📁 {}", name),
                format!("enterdir_{}_{}", action_type, i),
            )]);
        } else {
            let size = get_formatted_file_size(&path);
            let label = if is_multi {
                let mark = if selected_set.contains(&i) { "✅ " } else { "⬜️ " };
                format!("{}[{}] {}", mark, size, name)
            } else {
                format!("[{}] {}", size, name)
            };

            let cb_data = if is_multi {
                format!("toggle_{}_{}_{}", action_type, i, page)
            } else {
                format!("execsingle_{}_{}", action_type, i)
            };
            rows.push(vec![InlineKeyboardButton::callback(label, cb_data)]);
        }
    }

    let mut nav = Vec::new();
    if page > 0 {
        nav.push(InlineKeyboardButton::callback("⬅️ 上一页", format!("menu_{}_{}", action_type, page - 1)));
    }
    if page + 1 < total_pages {
        nav.push(InlineKeyboardButton::callback("➡️ 下一页", format!("menu_{}_{}", action_type, page + 1)));
    }
    if !nav.is_empty() { rows.push(nav); }

    if current_dir != state.config.base_dir {
        rows.push(vec![InlineKeyboardButton::callback("⬆️ 返回上一层目录", format!("updir_{}", action_type))]);
    }

    if is_multi && !selected_set.is_empty() {
        rows.push(vec![InlineKeyboardButton::callback(
            format!("▶️ 确认执行 (选中 {} 个文件)", selected_set.len()),
            format!("execbatch_{}", action_type),
        )]);
    }

    rows.push(vec![InlineKeyboardButton::callback("🔙 返回主菜单", "menu_main")]);

    let rel_path = current_dir.strip_prefix(&state.config.base_dir).unwrap_or(Path::new(""));
    let display_path = if rel_path.as_os_str().is_empty() { 
        "🏠".to_string() 
    } else { 
        format!("🏠/{}", rel_path.display()) 
    };
    
    let header = format!(
        "📂 路径: <code>{}</code>\n👉 模式: <b>[{}]</b> (页 {}/{})", 
        escape_html(&display_path), 
        escape_html(&action_type.to_uppercase()), 
        page + 1, 
        total_pages.max(1)
    );

    (header, InlineKeyboardMarkup::new(rows))
}